float64[3] qi   # previous eta
float64[3] pi   # previous epsilon